### ISSUE TITLE
Certificate decoding is overly liberal

### ISSUE THREAD

**Issue by paulidale** [openssl/openssl#21620]
When loading a certificate, OpenSSL accepts BER encodings.  According to the standards, all certificates should be strictly DER encoded.  This extra flexibility might result in a security issue at some stage.

The attached [`bad_cert.zip`](https://github.com/openssl/openssl/files/12234603/bad_cert.zip) archive contains three seemingly identical certificates (running the `check.sh` script produces identical output for each), yet all three files have different contents.
See the enclosed `README.md` for details.

Not that a simple change to always force DER decoding isn't sufficient, some portions of OpenSSL must be able to accept BER.

A suggestion was raised to have a global flags indicating string DER decoding by default.  This flag being switchable to allow BER.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 2 months otherwise.

**Comment by webhop123** [comment]
Hi, I am interested in working on this. Before I start, could you confirm that it is still open for implementation? My plan is to first map the certificate decoding paths that must remain BER-compatible, add regression tests for the sample certificates, and then introduce strict DER decoding for certificates with a controlled opt-in for the required BER cases. Is that the direction you want, or is there a preferred API or flag design?

**Comment by t8m** [comment]
Yes, this is still open. I'll assign it to you.

**Comment by webhop123** [comment]
Thanks @t8m for assigning this issue to me. Before I start, could you confirm whether this work has any funded bounty or sponsorship, and whether there is a preferred API or flag design for the BER opt-in?

**Comment by t8m** [comment]
No, unfortunately there is no funding behind this. As the first step of the work there should be a proposal for such API/flag.

**Comment by webhop123** [comment]
Proposed API/policy (corrected formatting; all public names are tentative):

1. Keep `ASN1_get_object()` and generic ASN.1 decoding BER-compatible. CMS, PKCS#7, PKCS#12, and other BER-capable outer structures must continue to work.
2. Mark the `X509` ASN.1 item as requiring definite, minimal DER length octets. Existing `d2i_X509()`-based paths then reject non-DER certificate lengths by default, including when an X509 object is nested in another ASN.1 structure.
3. Add an additive, per-call escape hatch without changing an existing function signature or public structure layout:

```c
#define ASN1_AFLG_REQUIRE_DER_LENGTH 16
#define ASN1_D2I_FLAG_ALLOW_BER_LENGTHS 0x1u

ASN1_VALUE *ASN1_item_d2i_ex2(ASN1_VALUE **val,
                               const unsigned char **in, long len,
                               const ASN1_ITEM *it,
                               OSSL_LIB_CTX *libctx, const char *propq,
                               unsigned int flags);

#define X509_D2I_FLAG_ALLOW_BER_LENGTHS \
        ASN1_D2I_FLAG_ALLOW_BER_LENGTHS

X509 *d2i_X509_ex2(X509 **x, const unsigned char **in, long len,
                    OSSL_LIB_CTX *libctx, const char *propq,
                    unsigned int flags);
```

Zero flags keep the strict X509 default. Unknown flag bits fail without consuming input. The opt-in is scoped to the top-level item passed to that call: passing it while decoding a BER container does not relax an independently marked embedded X509 object. This avoids process-global or mutable libctx policy and keeps the behavior thread-safe.

The proposed first scope is deliberately length-only:

- reject indefinite lengths;
- reject long-form lengths for values below 128;
- reject leading zero/redundant length octets;
- check the encoded children of constructed values;
- treat primitive contents, including extension OCTET STRING payloads, as opaque.

This resolves the malformed length encodings in the attached `b.pem` and `c.pem` without claiming complete DER canonicality. Tag-number minimality and other value-level DER rules are outside this first patch (for example, the separate concern in #28424).

Validation is performed on the original bytes at the marked ASN.1 item boundary after schema decoding. A decode/re-encode comparison is not reliable here because `X509_CINF` and `X509_NAME` can cache and reproduce the original non-canonical encoding.

I now have a local prototype of this design. Its regression coverage includes:

- issue fixture `a.pem` accepted by default;
- issue fixtures `b.pem` and `c.pem` rejected by default;
- `b.pem`, `c.pem`, and a derived indefinite-length certificate accepted only with the explicit opt-in, with their signatures verified against `ta.pem`;
- unknown flags, no cross-call flag leakage, X509 reuse with a new libctx/property query, and failure preserving the existing object/input pointer;
- nested marked-item policy, primitive opaque payloads, constructed children, and deep nesting;
- existing BER CMS, BER PKCS#12, and constructed-indefinite ASN.1 string compatibility.

The prototype passes a strict-warning build, `make doc-nits`, and the full test suite: 387 test files / 4579 tests, all successful.

Before I freeze ABI or open a PR, I would appreciate guidance on these policy points:

1. Do you prefer the generic `ASN1_item_d2i_ex2(..., flags)` plus X509 wrapper, or an X509-only public API?
2. Since there is no existing `d2i_X509_ex()`, should the first extended API be named `d2i_X509_ex()`, should both `_ex` and `_ex2` be added, or is `_ex2` acceptable here?
3. Is a raw-buffer opt-in sufficient for the first patch? With the top-level-only policy, existing PEM/BIO/fp, `d2i_X509_AUX()`, STORE, and embedded-container loading paths become strict with no flag-bearing compatibility entry point. I can add selected high-level opt-in variants if that recovery path is required.
4. Is definite/minimal length enforcement the desired phase-one scope for #21620, leaving the other DER canonicality rules to follow-up work?

I will keep the prototype local until the API/policy is agreed.
